Cleanup leftover PCIDevice methods from tt_silicon_driver

broskoTT opened the pull request:
A couple of methods were commented out with detecting hanged device disabled after #64 
This PR reintroduces it back, and moves all this PCIDevice related functions to that class.

vtangTT commented:
Originally commented these out as they didn't seem necessary and were in the critical read path. Are there plans to refactor or improve this functionality in the future?

broskoTT commented:
> Originally commented these out as they didn't seem necessary and were in the critical read path. Are there plans to refactor or improve this functionality in the future?

But this does detect a hang during read operation? Wouldn't commenting these out return wrong data?

joelsmithTT commented on device/pcie/pci_device.cpp:
```diff
@@ -493,3 +499,83 @@ void PCIDevice::write_tlb_reg(uint32_t byte_addr, std::uint64_t value_lower, std
+    return (scratch_data == 0xffffffffu);
+}
+
+bool PCIDevice::reset_by_sysfs() {
```
As far as I can tell, this is dead code.

joelsmithTT commented on device/pcie/pci_device.cpp:
```diff
@@ -493,3 +499,83 @@ void PCIDevice::write_tlb_reg(uint32_t byte_addr, std::uint64_t value_lower, std
+    return (reset_device.out.result == 0);
+}
+
+bool PCIDevice::auto_reset_board() {
```
I don't think UMD should be in the business of auto resetting anything.  UMD is not in a position to restore any the hardware to a known good state, so at best this is a convenience for the developer (won't have to run tt-smi).  At worst, it will complicate debugging efforts if user hangs a chip.

For what it's worth, I've never seen this work the way one would expect it to from reading the code.

broskoTT replied:
Makes sense to me. I'll remove the auto reset feature, we can bring it back if the need arises. I'll leave the hang detection though.

pjanevskiTT commented:
> But this does detect a hang during read operation? Wouldn't commenting these out return wrong data?

I found this useful few times while testing multiple things, for example on Blackhole this helped us debug few PCI hangs we had, so I would say we still need to keep this around. About that being on critical perf path, I agree, 99% of the time we are not going to hit reading 0xFF but we still check. I don't think we should remove this, we can easily remove it later on, I am not really sure what is the actual perf benefit of removing this. If we can measure it and then have an accurate numbers, that could be the argument to remove it.

Maybe we could expose, based on Release/debug build, safe debug read function, which is going to check for 0xFF always, and release read function which is not going to check that

joelsmithTT commented on device/pcie/pci_device.cpp:
```diff
@@ -493,3 +499,83 @@ void PCIDevice::write_tlb_reg(uint32_t byte_addr, std::uint64_t value_lower, std
+        std::uint32_t scratch_data = *get_register_address<std::uint32_t>(read_checking_offset);
+
+        if (auto_reset_board()) {
+            throw std::runtime_error("Read 0xffffffff from PCIE: auto-reset succeeded.");
+        } else {
+            throw std::runtime_error("Read 0xffffffff from PCIE: you should reset the board.");
```
What should the application do if it catches a `std::runtime_error`?  Unless it parses the `what()` string, the only reasonable thing to do is to terminate the program.

Either branch here is a fatal error.

broskoTT replied:
I think that is exactly the purpose of runtime_error. To tell the user this is fatal. 
Yes, both branches return runtime_error, as normal execution cannot continue after hang is detected. It is obvious from this code that the intention for auto reset feature was just to conveniently reset the card if it can, not try to continue normal execution.

joelsmithTT reviewed (approved):
Detecting a hung chip is fine, but I am not convinced that:
* the reset code actually works
* the reset code is something UMD should be doing

pjanevskiTT commented on device/pcie/pci_device.cpp:
```diff
@@ -493,3 +497,26 @@ void PCIDevice::write_tlb_reg(uint32_t byte_addr, std::uint64_t value_lower, std
 }
+
+bool PCIDevice::is_hardware_hung() {
+    volatile const void *addr = reinterpret_cast<const char *>(bar0_uc) + (get_architecture_implementation()->get_arc_reset_scratch_offset() + 6 * 4) - bar0_uc_offset;
```
nit: space between char and *

broskoTT replied:
I'm sorry, I specifically don't want to fix comments like since adding clang formatting to our repo will fix all of these automatically :) Will work on adding that soon.

broskoTT commented:
I've removed the auto reset code as @joelsmithTT had doubts on whether it was even functional (I also don't remember ever seeing it succeed).

@vtangTT and @pjanevskiTT raised concerns on the impact of detecting if the hardware is hung on reads, I've added this issue to further investigate that #156

pjanevskiTT commented on device/pcie/pci_device.cpp:
```diff
@@ -493,3 +497,26 @@ void PCIDevice::write_tlb_reg(uint32_t byte_addr, std::uint64_t value_lower, std
+    volatile const void *addr = reinterpret_cast<const char *>(bar0_uc) + (get_architecture_implementation()->get_arc_reset_scratch_offset() + 6 * 4) - bar0_uc_offset;
+    std::uint32_t scratch_data = *reinterpret_cast<const volatile std::uint32_t*>(addr);
+
+    return (scratch_data == 0xffffffffu);
```
Maybe we can put `0xffffffffu` to some const var or define, since we are using it in multiple places

broskoTT replied:
Added c_hang_read_value and also renamed detect_ffffffff_read to detect_hang_read

pjanevskiTT reviewed (approved):
nice, few nit comments
